Updated settings.py

manwinder123 opened the pull request:
Updated settings.py to use relative paths not absolute paths from feature_file_directories array. This allows me to have the following settings: 

```
{
    "feature_file_directories":
    [
      "features"
    ]
}
```

that give me the ability to work on different projects while not changing the settings file. Before I would have to change the absolute path inside the settings file when i switched projects.

manwinder123 committed:
Updated settings.py

Updated settings.py to use relative paths not absolute paths from feature_file_directories array. This allows me to have the following settings: 
```
{
    "feature_file_directories":
    [
      "features"
    ]
}
```
that give me the ability to work on different projects while not changing the settings file. Before I would have to change the absolute path inside the settings file when i switched projects.

austincrft commented:
This would remove the ability for people to include steps from other projects, which some people use.

What if we left the existing option, but also supplied an option to pattern-match directory names with regex? This would allow for both features.

manwinder123 commented:
I guess that would work. But what if we added another array in the user settings for absolute paths. Those would be added to the list easily, pretty much the few lines of code you had before. So you get the standard features path inside the project folder as well as any absolute paths.

manwinder123 committed:
re-added absolute paths to user settings

This allows the user to include absolute paths as well as the relative paths (what this pr was about)

austincrft commented on gherkin_auto_complete_plus/utilities/settings.py:
```diff
@@ -21,7 +21,21 @@ def get_feature_directories():
+    feature_directories_paths = feature_directories_absolute_path
+
+    project_data = sublime.active_window().project_data()
+    project_folder = project_data['folders'][0]['path']    
```
Couldn't this be written using the sublime.Window.folders method?

``` python
sublime.active_window().folders()
```

manwinder123 replied:
@austincrft yup, i'll change it

austincrft commented:
I still think this should be done with pattern matching. The user could supply a list of patters similar to the way a `.gitignore` is done.

So a user would specify

``` javascript
{
    "patterns":
    [
        "*/features",
        "*/feature-files"
    ]
}
```

and the return for `get_feature_directories` would be a list of those pattern matched paths. Like so

``` python
['/src/sublime-gherkin-auto-complete-plus/*/features', '/src/sublime-gherkin-auto-complete-plus/*/feature-files']
```

There would have to be testing around the performance of this feature, though. Finding that pattern would still be a recursive search.

manwinder123 committed:
updated settings.py

changed the way the project folders are retrieved.

manwinder123 commented:
@austincrft would the pattern matching be just for the project folder? If so how would the user include feature folders from other projects?

austincrft commented:
We would just append the pattern matching onto the project folder and search for that.

```
'/project/folder/path/{pattern goes here}
```

This will require quite a bit of code changes. If you'd prefer, you can use your local workaround and I can work on a more fully-fledged feature.

manwinder123 commented:
I'll look into this. I was thinking using OS.walk, I would go through the folder names and see if the regex matches. 

Since I am only going through project folders how would I be able to go to a different project (thats not in sublime) folder and return its feature path. I was thinking I would do a secondary check after the project folder search, I would see if any of the regex patterns provided were also valid paths.

Also wouldn't adding the path and then the pattern kinda take away from what we are trying to do, if you know the path then just put the path, no need to put the pattern.
`/project/folder/path/{pattern goes here}`

`/project/folder/path/features-folder`

manwinder123 committed:
Updated settings.pyt

manwinder123 commented:
I have changed the get_feature_directories function so that you can put absolute paths as well as relative paths. 

```
{
    "feature_file_directories":
    [
        "/Users/userfifteen/Projects/Project-fourteen/features",
        "/features"
    ]
}
```

austincrft commented:
That's actually what I'm doing in the `get_feature_files` method in [gherkin_parser.py](https://github.com/austincrft/sublime-gherkin-auto-complete-plus/blob/master/gherkin_auto_complete_plus/utilities/gherkin_parser.py). I'm using `os.walk` and `fnmatch.filter` to recursively check every file in the project. The performance isn't great for large projects, so it's very possible that the pattern-matching will cause the same problems.

The `feature_file_directories` option already uses absolute paths, so you wouldn't need to change anything to that.

The object is to eliminate _per-project_ settings, because that is difficult to setup in Sublime Text. If you open a project that doesn't have the _exact relative path_ `/features`, then you'd have to modify the options for that project. If we're doing pattern-matching for the directory name `features`, then it can be anywhere in the file tree and still pick it up. Plus, using `fnmatch` would allow people to match things like `*/*feature*` which would match directories named `features` and `feature-files`.

manwinder123 commented:
I understand the objective. 

The code I have added doesn't use regex but it does look for the `features` (assuming the settings has `features` in it) in the array of folders in the project, it doesn't need an exact match (though the text has to be exact), it matches for example `/users/test/project/features`. 

I could step the search up a notch and search all folders inside the project. It currently searches the folders in the root of the projects.

austincrft commented:
I'm going to close this for now. If there is more interest in something like this, I'll work towards implementing it.
